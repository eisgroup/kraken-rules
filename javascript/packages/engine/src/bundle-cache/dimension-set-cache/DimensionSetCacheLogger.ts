export type DimensionSetCacheLogger = { logWarning: (message: string) => void }
