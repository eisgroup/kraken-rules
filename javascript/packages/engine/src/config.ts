/*
 *  Copyright 2020 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { debug } from './debugger/Debugger'

/**
 * Configures Kraken evaluation. It can be configured from the console.
 * Open the console and  modify the config options
 * @example
 * global.Kraken.debug = false;
 * global.Kraken.info = false;
 * global.Kraken.debug = false;
 * @see Logger
 * @since 1.1.1
 */
export interface KrakenConfig {
    logger: {
        enabled: boolean
        debug: boolean
        info: boolean
        error: boolean
        warning: boolean
        entriesCount: number
        logs: string[]
        getLogs(): void
        clear(): void
    }
    debugger: debug.api.Debugger
}
export type WithKraken = {
    Kraken: KrakenConfig
}

function init(): void {
    const logger: KrakenConfig['logger'] = {
        enabled: process.env.NODE_ENV !== 'production',
        debug: false,
        info: true,
        error: true,
        warning: true,
        entriesCount: 2000,
        logs: [],
        clear(): void {
            const config = (global as unknown as WithKraken).Kraken
            if (config.logger.logs) {
                config.logger.logs = []
            }
        },
        getLogs(): void {
            const config = (global as unknown as WithKraken).Kraken
            // applicable only for the browser
            if (config.logger.logs && document && window) {
                const blob = new Blob([config.logger.logs.join('\n')], { type: 'text/plain' })
                const url = window.URL.createObjectURL(blob)
                const a = document.createElement('a')
                a.style.display = 'none'
                a.href = url
                a.download = `kraken-logs-${new Date().getTime()}.txt`
                document.body.appendChild(a)
                a.click()
                window.URL.revokeObjectURL(url)
                document.body.removeChild(a)
            }
        },
    }
    if (global) {
        ;(global as unknown as WithKraken).Kraken = {
            logger,
            debugger: new debug.impl.DevToolsDebugger(),
        }
        return
    }
}

function getConfig(): KrakenConfig {
    if (global) {
        return (global as unknown as { Kraken: KrakenConfig })['Kraken']
    }
    throw new Error('Kraken configuration was not initialized')
}

export const krakenConfig = { init, getConfig }
