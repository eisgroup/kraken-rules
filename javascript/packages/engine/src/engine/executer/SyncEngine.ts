/*
 *  Copyright 2018 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { ContextInstanceInfoResolver } from '../contexts/info/ContextInstanceInfoResolver'
import { DataObjectInfoResolver } from '../contexts/info/DataObjectInfoResolver'
import { EntryPointResult } from 'kraken-engine-api'
import { ExecutionSession } from '../ExecutionSession'
import { requireDefinedValue } from '../../utils/Utils'
import { ContextModelTree } from '../../models/ContextModelTree'
import { RulePayloadProcessor } from '../RulePayloadProcessor'
import { ExpressionEvaluator, KelFunction } from '../runtime/expressions/ExpressionEvaluator'
import { DataContextBuilder } from '../contexts/data/DataContextBuilder'
import { ExtractedChildDataContextBuilder } from '../contexts/data/ExtractedChildDataContextBuilder'
import { ContextDataExtractorImpl } from '../contexts/data/extraction/ContextDataExtractorImpl'
import { FunctionRegistry } from '../runtime/expressions/functionLibrary/Registry'
import { RuleOverrideContextExtractorImpl } from '../results/RuleOverrideContextExtractor'
import { logger } from '../../utils/DevelopmentLogger'
import { DataContextUpdaterImpl } from '../contexts/data/updater/DataContextUpdaterImpl'
import { ContextDataExtractor } from '../contexts/data/extraction/ContextDataExtractor.types'
import { CachingContextDataExtractor } from '../contexts/data/extraction/CachingContextDataExtractor'
import { ReferencePathResolverImpl } from '../contexts/ccr/ReferencePathResolverImpl'
import { PathCardinalityResolver } from '../contexts/ccr/PathCardinalityResolver'
import { CommonPathResolver } from '../contexts/ccr/CommonPathResolver'
import { CachingReferencePathResolver } from '../contexts/ccr/CachingReferencePathResolver'
import { ENGINE_VERSION } from '../generated/engineVersion'
import { DefaultEntryPointResult } from '../../dto/DefaultEntryPointResult'
import { ErrorCode, KrakenRuntimeError } from '../../error/KrakenRuntimeError'
import { EntryPointBundle } from '../../models/EntryPointBundle'
import EntryPointEvaluation = EntryPointBundle.EntryPointEvaluation
import { EvaluationMode, isSupportedPayloadType } from '../runtime/EvaluationMode'
import { EntryPointBundleCache } from '../../bundle-cache/EntryPointBundleCache'
import { OrderedEvaluationLoop } from '../processors/OrderedEvaluationLoop'

export interface EvaluationConfig {
    context: {
        /**
         * Dimension values, that will be as a filter to get rules and entrypoints
         * that applies to that dimensions
         */
        dimensions?: Record<string, unknown>
        externalData?: Record<string, unknown>
        [keyof: string]: unknown
    }

    /**
     * Currency to use when interpreting monetary types
     */
    currencyCd: string

    /**
     * Evaluation mode to use when evaluating rules. Evaluation mode dictates what type
     * of rules will be evaluated during a Kraken engine call.
     *
     * If evaluation mode is not specified, then {@link EvaluationMode.ALL} is used by default.
     */
    evaluationMode?: EvaluationMode

    /**
     * This id enables Kraken to treat different rules evaluations as on evaluation.
     * This enables data extraction caching between different evaluations.
     */
    evaluationId?: string
}

/**
 * Config to pass when calling {@link Engine.withConfig}
 */
export interface SyncEngineConfig {
    /**
     * Cached artifacts grouped by dimensions
     */
    cache: EntryPointBundleCache
    /**
     * Resolves context definition name for specified data object instance
     */
    dataInfoResolver: DataObjectInfoResolver
    /**
     * Provides three different info resolution methods, which are called depending on how
     * particular context instance data object was extracted
     */
    contextInstanceInfoResolver: ContextInstanceInfoResolver<unknown>

    /**
     * Static context definition model
     */
    modelTree: ContextModelTree.ContextModelTree

    /**
     * Custom functions that can be invoked in rules within this engine instance
     */
    functions?: KelFunction[]

    /**
     * Used for compatibility check with the backend.
     * {@link EntryPointBundle.EntryPointBundle}#engineVersion must match against
     * this version. Default version is backend artifact version at a release time.
     */
    engineCompatibilityVersion?: string
}
type CachedDataExtractionServices = {
    contextDataExtractor: CachingContextDataExtractor
}

export class SyncEngine {
    readonly #rulePayloadProcessor: RulePayloadProcessor
    readonly #dataContextBuilder: DataContextBuilder
    readonly #contextDataExtractor: ContextDataExtractor
    readonly #bundleCache: EntryPointBundleCache
    readonly #dataInfoResolver: DataObjectInfoResolver

    readonly #expressionEvaluator: ExpressionEvaluator
    readonly #engineCompatibilityVersion?: string

    private evaluationId?: string
    private serviceCache?: CachedDataExtractionServices

    constructor(config: SyncEngineConfig) {
        function resolveInheritance(name: string): string[] {
            return config.modelTree.contexts[name].inheritedContexts
        }
        this.#expressionEvaluator = new ExpressionEvaluator(
            FunctionRegistry.createInstanceFunctions(config.dataInfoResolver, resolveInheritance),
            config.functions ?? [],
        )
        this.#dataContextBuilder = new DataContextBuilder(config.modelTree, config.contextInstanceInfoResolver)
        this.#contextDataExtractor = new ContextDataExtractorImpl(
            config.modelTree,
            new ExtractedChildDataContextBuilder(this.#dataContextBuilder, this.#expressionEvaluator),
        )
        const referencePathResolver = new CachingReferencePathResolver(
            new ReferencePathResolverImpl(
                config.modelTree.pathsToNodes,
                new PathCardinalityResolver(config.modelTree.contexts),
                new CommonPathResolver(),
            ),
        )
        const dataContextUpdater = new DataContextUpdaterImpl(referencePathResolver, (root, path) =>
            this.#contextDataExtractor.extractByPath(root, path),
        )
        this.#rulePayloadProcessor = new RulePayloadProcessor(
            this.#expressionEvaluator,
            new RuleOverrideContextExtractorImpl(),
            dataContextUpdater,
        )
        this.#bundleCache = config.cache
        this.#dataInfoResolver = config.dataInfoResolver
        this.#engineCompatibilityVersion = config.engineCompatibilityVersion
    }

    evaluate(data: object, entryPointName: string, config: EvaluationConfig): EntryPointResult {
        return logger.group(
            () => `Kraken logs: '${entryPointName}'`,
            () => this.doEvaluate(data, entryPointName, config),
            result => `Kraken logs: '${entryPointName}'. Timestamp: ${result.evaluationTimestamp.toUTCString()}`,
        )
    }

    evaluateSubTree(data: object, node: object, entryPointName: string, config: EvaluationConfig): EntryPointResult {
        return logger.group(
            () => `Kraken logs: '${entryPointName}'`,
            () => this.doEvaluate(data, entryPointName, config, this.requireValidNode(node)),
            result => `Kraken logs: '${entryPointName}'. Timestamp: ${result.evaluationTimestamp.toUTCString()}`,
        )
    }

    private doEvaluate = (data: object, entryPointName: string, evaluationConfig: EvaluationConfig, node?: object) => {
        const start = Date.now()
        logger.info(() => {
            return { evaluationConfig }
        })
        if (node) {
            const contextName = this.#dataInfoResolver.resolveName(node)
            const id = this.#dataInfoResolver.resolveId(node)
            logger.info(() => `With restriction node: ${contextName}:${id}`)
        }
        const bundle = this.#bundleCache.get(entryPointName, evaluationConfig.context.dimensions || {})

        const version = this.#engineCompatibilityVersion || ENGINE_VERSION
        if (bundle.engineVersion && this.normalizedVersion(bundle.engineVersion) !== this.normalizedVersion(version)) {
            logger.warning(
                () => `UI engine (${version}) and backend engine (${bundle.engineVersion}) versions are different.`,
            )
        }

        const filteredRulesEvaluation = this.filterRules(bundle.evaluation, evaluationConfig.evaluationMode)

        if (!filteredRulesEvaluation.rules.length) {
            logger.debug(() => `Kraken rules evaluation took ${Math.round(Date.now() - start)} ms`, true)
            return DefaultEntryPointResult.empty()
        }

        this.#expressionEvaluator.rebuildFunctions() //
        requireDefinedValue(evaluationConfig.currencyCd, 'Currency code cannot be absent')
        const session = new ExecutionSession(evaluationConfig, bundle.expressionContext)
        const cachingServices = this.getCachingServices(evaluationConfig.evaluationId)
        const result = OrderedEvaluationLoop.getInstance({
            rulePayloadProcessor: this.#rulePayloadProcessor,
            contextBuilder: this.#dataContextBuilder,
            contextDataExtractor: cachingServices.contextDataExtractor,
            restriction: node,
        }).evaluate(filteredRulesEvaluation, data, session)
        logger.debug(() => `Kraken rules evaluation took ${Math.round(Date.now() - start)} ms`, true)
        return result
    }

    private filterRules(
        entryPointEvaluation: EntryPointEvaluation,
        evaluationMode?: EvaluationMode,
    ): EntryPointEvaluation {
        if (evaluationMode === undefined || evaluationMode === EvaluationMode.ALL) {
            return entryPointEvaluation
        }

        return {
            ...entryPointEvaluation,
            rules: entryPointEvaluation.rules.filter(value =>
                isSupportedPayloadType(evaluationMode, value.payload.type),
            ),
        }
    }

    /**
     * Normalizes stage version postfix to be compatible between mvn snapshot version and semver.
     * Theoretically, this could create false positive in some extremely rare case (1.0.0-12.3 === 1.0.0-1.23),
     * but this is unlikely to happen in practice
     *
     * @param version
     * @private
     */
    private normalizedVersion(version: string): string {
        const versionParts = version.split('-')
        if (versionParts.length !== 2) {
            return version
        }

        const normalizedSnapshotPart = versionParts[1].replace('.', '')
        return `${versionParts[0]}-${normalizedSnapshotPart}`
    }

    private getCachingServices(evaluationId?: string): CachedDataExtractionServices {
        if (evaluationId) {
            if (this.evaluationId === evaluationId && this.serviceCache) {
                // present by this evaluation id
                logger.debug(() => 'Using Cached Data Extraction Services from previous execution')
                return this.serviceCache
            } else {
                // absent by this evaluation id
                logger.debug(() => 'Initializing new Cached Data Extraction Services')
                this.evaluationId = evaluationId
                this.serviceCache = { contextDataExtractor: this.getContextDataExtractor() }
                return this.serviceCache
            }
        }
        // no evaluation id
        this.evaluationId = undefined
        this.serviceCache = undefined
        return { contextDataExtractor: this.getContextDataExtractor() }
    }

    private getContextDataExtractor(): CachingContextDataExtractor {
        return new CachingContextDataExtractor(this.#contextDataExtractor)
    }

    private requireValidNode = (node: object): object => {
        requireDefinedValue(node, 'Restriction node cannot be absent')
        const validationErrors = this.#dataInfoResolver.validate(node)
        if (validationErrors.length) {
            throw new KrakenRuntimeError(
                ErrorCode.INCORRECT_CONTEXT_INSTANCE,
                'Restriction node is invalid Context instance:\n\t' + validationErrors.map(x => x.message).join('\n\t'),
            )
        }
        return node
    }
}
