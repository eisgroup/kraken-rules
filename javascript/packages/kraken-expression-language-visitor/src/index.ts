export { createParser } from "./parse";
export { KelVisitor } from "./antlr/generated/KelVisitor";
export { SyntaxError } from "./KrakenParser";
export * from "./antlr/generated/Kel";

export { ParserRuleContext } from "antlr4ts";
export { AbstractParseTreeVisitor } from "antlr4ts/tree/AbstractParseTreeVisitor";
export { ErrorNode } from "antlr4ts/tree/ErrorNode";
export { Interval } from "antlr4ts/misc/Interval";
