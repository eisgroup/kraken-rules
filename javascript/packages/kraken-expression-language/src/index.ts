export * from './scope/Scope'
export * from './scope/ScopeType'
export * from './visitor/Node'
export * from './visitor/AstGeneratingVisitor'
export * from './visitor/NodeUtils'

export { KelParser, SyntaxError } from 'kraken-expression-language-visitor'

export * from './visitor/autocomplete/CompletionItem'
export * from './visitor/type-at-location/LocationInfo'

export * from './public/Cache'
export * from './public/KelService'
export * from './public/ScopeDeserializer'

export { Type } from './type/Types'
