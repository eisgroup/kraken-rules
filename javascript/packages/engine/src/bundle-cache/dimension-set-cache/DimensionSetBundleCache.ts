import { Rule } from 'kraken-model'
import {
    SystemMessageBuilder,
    KrakenRuntimeError,
    NO_BUNDLE_CACHE_BY_ENTRYPOINT_AND_DIMENSIONS,
    NO_BUNDLE_CACHE_BY_ENTRYPOINT,
} from '../../error/KrakenRuntimeError'
import { EntryPointBundle } from '../../models/EntryPointBundle'
import { DimensionSetCache } from './DimensionSetCache'
import { ExpressionContext, ExpressionContextManager } from '../expression-context-manager/ExpressionContextManager'
import { Dimensions, DimensionSet, EntryPointBundleCache } from '../EntryPointBundleCache'
import { Cache } from '../Cache'
import { DimensionSetCacheLogger } from './DimensionSetCacheLogger'
import { Reducer as r } from 'declarative-js'

type EntryPointName = string

export class DimensionSetBundleCache implements EntryPointBundleCache, ExpressionContextManager, Cache {
    constructor(logger: DimensionSetCacheLogger, expressionContextManager: ExpressionContextManager) {
        this.#logger = logger
        this.#expressionContextManager = expressionContextManager
    }

    #bundleCache = new BundleCache()
    #rulesCache = new Map<EntryPointName, DimensionSetCache>()
    #logger: DimensionSetCacheLogger
    #expressionContextManager: ExpressionContextManager

    clearCache(): void {
        this.#expressionContextManager.clearExpressionContext()
        this.#bundleCache.clear()
        this.#rulesCache.clear()
    }

    add(entryPointName: string, dimensions: Dimensions, bundle: EntryPointBundle.EntryPointBundle): void {
        const entryPointCache = this.#rulesCache.get(entryPointName) ?? new DimensionSetCache(this.#logger)
        entryPointCache.add(dimensions, bundle.evaluation.rules)
        this.#rulesCache.set(entryPointName, entryPointCache)
        this.#bundleCache.set(entryPointName, bundle, dimensions)
    }
    get(entryPointName: string, dimensions: Dimensions): EntryPointBundle.EntryPointBundle {
        const entryPointCache = this.#rulesCache.get(entryPointName)
        if (!entryPointCache) {
            const m = new SystemMessageBuilder(NO_BUNDLE_CACHE_BY_ENTRYPOINT).parameters(entryPointName).build()
            throw new KrakenRuntimeError(m)
        }

        if (!entryPointCache.isCached(dimensions)) {
            const m = new SystemMessageBuilder(NO_BUNDLE_CACHE_BY_ENTRYPOINT_AND_DIMENSIONS)
                .parameters(entryPointName, dimensions)
                .build()
            throw new KrakenRuntimeError(m)
        }

        const rules = entryPointCache.get(dimensions)
        const expressionContext = this.#expressionContextManager.getExpressionContext()
        return this.#bundleCache.get(entryPointName, rules, expressionContext, dimensions)
    }
    calculateExcludes(entryPointName: string, dimensions: Dimensions): DimensionSet[] {
        const cache = this.#rulesCache.get(entryPointName)
        if (!cache) {
            return []
        }
        return cache.calculateExcludes(dimensions)
    }
    isCached(entryPointName: string, dimensions: Dimensions): boolean {
        return (
            this.#bundleCache.has(entryPointName, dimensions) &&
            this.#rulesCache.has(entryPointName) &&
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            this.#rulesCache.get(entryPointName)!.isCached(dimensions)
        )
    }
    clearRulesCache(): void {
        this.#bundleCache.clear()
        for (const cache of this.#rulesCache.values()) {
            cache.clear()
        }
        this.#rulesCache.clear()
    }

    setExpressionContext(expressionContext: Record<string, unknown>): void {
        this.#expressionContextManager.setExpressionContext(expressionContext)
    }
    clearExpressionContext(): void {
        this.#expressionContextManager.clearExpressionContext()
    }
    isExpressionContextPresent(): boolean {
        return this.#expressionContextManager.isExpressionContextPresent()
    }
    getExpressionContext(): ExpressionContext {
        return this.#expressionContextManager.getExpressionContext()
    }
}

class BundleCache {
    // key is: entrypoint name-sorted dimension values
    private cache = new Map<string, EntryPointBundle.EntryPointBundle>()

    set(entryPointName: string, bundle: EntryPointBundle.EntryPointBundle, dimensions: Record<string, unknown>): void {
        this.cache.set(this.getKey(entryPointName, dimensions), bundle)
    }

    get(
        entryPointName: string,
        rules: Rule[],
        expressionContext: ExpressionContext,
        dimensions: Record<string, unknown>,
    ): EntryPointBundle.EntryPointBundle {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const bundle = this.cache.get(this.getKey(entryPointName, dimensions))!
        return {
            ...bundle,
            expressionContext,
            evaluation: {
                ...bundle.evaluation,
                rules,
            },
        }
    }

    has(entryPointName: string, dimensions: Record<string, unknown>): boolean {
        return this.cache.has(this.getKey(entryPointName, dimensions))
    }
    clear(): void {
        this.cache.clear()
    }

    private getKey(entryPointName: string, dimensions: Record<string, unknown>): string {
        const allDimensionKeys = Object.keys(dimensions)
        const sorted = allDimensionKeys.sort().reduce(
            r.toObject(
                x => x,
                x => dimensions[x],
            ),
            {},
        )
        return `${entryPointName}-${JSON.stringify(sorted)}`
    }
}
