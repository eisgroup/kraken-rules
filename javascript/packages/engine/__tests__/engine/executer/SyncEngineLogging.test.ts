/*
 *  Copyright 2023 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
import { SyncEngine } from '../../../src/engine/executer/SyncEngine'
import { mock } from '../../mock'
import { PayloadBuilder, RulesBuilder } from 'kraken-model-builder'
import { Payloads } from 'kraken-model'
import UsageType = Payloads.Validation.UsageType
import { DimensionSetBundleCache } from '../../../src/bundle-cache/dimension-set-cache/DimensionSetBundleCache'
import { ExpressionContextManagerImpl } from '../../../src'
import { KrakenConfig } from '../../../src/config'

type ThisWithKraken = typeof global & { Kraken: KrakenConfig }

let engine: SyncEngine

const Policy = mock.modelTreeJson.contexts.Policy

let previousEnvironment

beforeEach(() => {
    ;(global as ThisWithKraken).Kraken.logger.clear()
    previousEnvironment = process.env.NODE_ENV

    const cache = new DimensionSetBundleCache({ logWarning: () => void 0 }, new ExpressionContextManagerImpl())
    cache.setExpressionContext({})

    cache.add(
        'EngineDebugLoggingTest',
        {},
        {
            evaluation: {
                delta: false,
                fieldOrder: [`${Policy.name}.${Policy.fields.policyNumber.name}`],
                rules: [
                    new RulesBuilder()
                        .setContext(Policy.name)
                        .setName('Accessibility')
                        .setTargetPath(Policy.fields.policyNumber.name)
                        .setPayload(PayloadBuilder.accessibility().notAccessible())
                        .build(),
                    new RulesBuilder()
                        .setContext(Policy.name)
                        .setName('Visibility')
                        .setTargetPath(Policy.fields.policyNumber.name)
                        .setPayload(PayloadBuilder.visibility().notVisible())
                        .build(),
                    new RulesBuilder()
                        .setContext(Policy.name)
                        .setName('Usage')
                        .setTargetPath(Policy.fields.policyNumber.name)
                        .setPayload(PayloadBuilder.usage().is(UsageType.mandatory))
                        .build(),
                    new RulesBuilder()
                        .setContext(Policy.name)
                        .setName('Default')
                        .setTargetPath(Policy.fields.policyNumber.name)
                        .setPayload(PayloadBuilder.default().to('"P0001"'))
                        .build(),
                    new RulesBuilder()
                        .setContext(Policy.name)
                        .setName('Assert')
                        .setTargetPath(Policy.fields.policyNumber.name)
                        .setCondition('!this._eq(__references__.Policy.policyNumber, undefined)', [
                            { name: 'Policy', type: 'CROSS_CONTEXT' },
                        ])
                        .setPayload(PayloadBuilder.asserts().that('__dataObject__.policyNumber == "P0001"'))
                        .addDependency({ contextName: 'Policy', ccrDependency: true, selfDependency: false })
                        .build(),
                    new RulesBuilder()
                        .setContext(Policy.name)
                        .setName('RegExp')
                        .setTargetPath(Policy.fields.policyNumber.name)
                        .setPayload(PayloadBuilder.regExp().match('[A-Z]+'))
                        .build(),
                    new RulesBuilder()
                        .setContext(Policy.name)
                        .setName('Length')
                        .setTargetPath(Policy.fields.policyNumber.name)
                        .setPayload(PayloadBuilder.lengthLimit().limit(5))
                        .build(),
                    new RulesBuilder()
                        .setContext(Policy.name)
                        .setName('MinSize')
                        .setTargetPath(Policy.fields.riskItems.name)
                        .setPayload(PayloadBuilder.size().min(0))
                        .build(),
                    new RulesBuilder()
                        .setContext(Policy.name)
                        .setName('MaxSize')
                        .setTargetPath(Policy.fields.riskItems.name)
                        .setPayload(PayloadBuilder.size().max(5))
                        .build(),
                    new RulesBuilder()
                        .setContext(Policy.name)
                        .setName('SizeRange')
                        .setTargetPath(Policy.fields.riskItems.name)
                        .setPayload(PayloadBuilder.size().range(0, 1))
                        .build(),
                    new RulesBuilder()
                        .setContext(Policy.name)
                        .setName('NumberSet')
                        .setTargetPath(Policy.fields.termNo.name)
                        .setPayload(PayloadBuilder.numberSet().greaterThanOrEqualTo(1, 1))
                        .build(),
                    new RulesBuilder()
                        .setContext(Policy.name)
                        .setName('ValueList')
                        .setTargetPath(Policy.fields.contractTermTypeCd.name)
                        .setPayload(
                            PayloadBuilder.valueList().valueList({
                                values: ['MONTHLY', 'ANNUAL'],
                                valueType: 'STRING',
                            }),
                        )
                        .build(),
                ],
                entryPointName: 'EngineDebugLoggingTest',
            },
            engineVersion: '1',
            expressionContext: {},
        },
    )

    jest.useFakeTimers('modern').setSystemTime(new Date('2022-01-01T10:00:00Z'))

    engine = new SyncEngine({
        cache: cache,
        dataInfoResolver: mock.spi.dataResolver,
        contextInstanceInfoResolver: mock.spi.instance,
        modelTree: mock.modelTree,
        functions: mock.policyFunctions,
        engineCompatibilityVersion: '1.2.3',
    })
})

afterEach(() => {
    ;(global as ThisWithKraken).Kraken.logger.clear()
    process.env.NODE_ENV = previousEnvironment

    jest.useRealTimers()
})

describe('SyncEngine', () => {
    it('Should debug log in development environment', () => {
        ;(global as ThisWithKraken).Kraken.logger.debug = true
        process.env.NODE_ENV = 'development'

        const data = mock.data.dataContextCustom({
            riskItems: [
                {
                    model: 'WV',
                },
            ],
        })

        engine.evaluate(data.dataObject, 'EngineDebugLoggingTest', { ...mock.evaluationConfig })

        const logs = (global as ThisWithKraken).Kraken.logger.logs

        expect(logs.join('\n')).toMatchSnapshot()
    })
})
