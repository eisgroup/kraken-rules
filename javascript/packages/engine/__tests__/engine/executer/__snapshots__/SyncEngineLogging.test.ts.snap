// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SyncEngine Should debug log in development environment 1`] = `
"-> Kraken logs: 'EngineDebugLoggingTest'
   -- {
        \\"evaluationConfig\\": {
          \\"context\\": {},
          \\"currencyCd\\": \\"USD\\"
        }
      }
   -- UI engine (1.2.3) and backend engine (1) versions are different.
   -> Evaluating default rules
      -> Context extraction
         -- Extracting context definition 'Policy' instance
         -- Path to extract 'Policy'
         -- Resolved 1 data context(s) for rule 'Default' target 'Policy':
            Policy:0
      <- Context extraction
      -> Evaluating rule 'Default' on Policy:0:policyNumber
         -- Evaluating expression '\\"P0001\\"'. 
         -- Evaluated 'DEFAULT'. Before value: ''. After value: 'P0001'.
      <- Evaluated rule 'Default' on Policy:0:policyNumber. Evaluation status - APPLIED.
   <- Evaluated default rules.
      Evaluated rule 'Default' on a total of 1 instances.
   -> Evaluating rule 'Accessibility'
      -- Resolved 1 data context(s) for rule 'Accessibility' target 'Policy':
         Policy:0
      -> Evaluating rule 'Accessibility' on Policy:0:policyNumber
         -- Evaluated 'ACCESSIBILITY'. The field is set to be disabled.
      <- Evaluated rule 'Accessibility' on Policy:0:policyNumber. Evaluation status - APPLIED.
   <- Evaluated rule 'Accessibility' on a total of 1 instances.
   -> Evaluating rule 'Visibility'
      -- Resolved 1 data context(s) for rule 'Visibility' target 'Policy':
         Policy:0
      -> Evaluating rule 'Visibility' on Policy:0:policyNumber
         -- Evaluated 'VISIBILITY'. The field is set to be hidden.
      <- Evaluated rule 'Visibility' on Policy:0:policyNumber. Evaluation status - APPLIED.
   <- Evaluated rule 'Visibility' on a total of 1 instances.
   -> Evaluating rule 'Usage'
      -- Resolved 1 data context(s) for rule 'Usage' target 'Policy':
         Policy:0
      -> Evaluating rule 'Usage' on Policy:0:policyNumber
         -- Evaluated 'USAGE' to true. Field is mandatory and it has value 'P0001'.
      <- Evaluated rule 'Usage' on Policy:0:policyNumber. Evaluation status - APPLIED.
   <- Evaluated rule 'Usage' on a total of 1 instances.
   -> Evaluating rule 'Assert'
      -- Resolved 1 data context(s) for rule 'Assert' target 'Policy':
         Policy:0
      -> Evaluating rule 'Assert' on Policy:0:policyNumber
         -- Evaluating expression '!this._eq(__references__.Policy.policyNumber, undefined)'. Cross context references:
            Policy=Policy:0.
         -- Evaluating expression '__dataObject__.policyNumber == \\"P0001\\"'. 
         -- Evaluated 'ASSERTION' expression '__dataObject__.policyNumber == \\"P0001\\"' to true.
      <- Evaluated rule 'Assert' on Policy:0:policyNumber. Evaluation status - APPLIED.
   <- Evaluated rule 'Assert' on a total of 1 instances.
   -> Evaluating rule 'RegExp'
      -- Resolved 1 data context(s) for rule 'RegExp' target 'Policy':
         Policy:0
      -> Evaluating rule 'RegExp' on Policy:0:policyNumber
         -- Evaluated 'REGEX' to true. Value 'P0001' matches regular expression '[A-Z]+'.
      <- Evaluated rule 'RegExp' on Policy:0:policyNumber. Evaluation status - APPLIED.
   <- Evaluated rule 'RegExp' on a total of 1 instances.
   -> Evaluating rule 'Length'
      -- Resolved 1 data context(s) for rule 'Length' target 'Policy':
         Policy:0
      -> Evaluating rule 'Length' on Policy:0:policyNumber
         -- Evaluated 'LENGTH' to true. Expected length '5'. Actual length '5'
      <- Evaluated rule 'Length' on Policy:0:policyNumber. Evaluation status - APPLIED.
   <- Evaluated rule 'Length' on a total of 1 instances.
   -> Evaluating rule 'MinSize'
      -- Resolved 1 data context(s) for rule 'MinSize' target 'Policy':
         Policy:0
      -> Evaluating rule 'MinSize' on Policy:0:riskItems
         -- Evaluated 'SIZE' to true. Expected size no less than 0 - actual size is 1.
      <- Evaluated rule 'MinSize' on Policy:0:riskItems. Evaluation status - APPLIED.
   <- Evaluated rule 'MinSize' on a total of 1 instances.
   -> Evaluating rule 'MaxSize'
      -- Resolved 1 data context(s) for rule 'MaxSize' target 'Policy':
         Policy:0
      -> Evaluating rule 'MaxSize' on Policy:0:riskItems
         -- Evaluated 'SIZE' to true. Expected size no more than 5 - actual size is 1.
      <- Evaluated rule 'MaxSize' on Policy:0:riskItems. Evaluation status - APPLIED.
   <- Evaluated rule 'MaxSize' on a total of 1 instances.
   -> Evaluating rule 'SizeRange'
      -- Resolved 1 data context(s) for rule 'SizeRange' target 'Policy':
         Policy:0
      -> Evaluating rule 'SizeRange' on Policy:0:riskItems
         -- Evaluated 'SIZE_RANGE' to true. Collection field size is within expected range..
      <- Evaluated rule 'SizeRange' on Policy:0:riskItems. Evaluation status - APPLIED.
   <- Evaluated rule 'SizeRange' on a total of 1 instances.
   -> Evaluating rule 'NumberSet'
      -- Resolved 1 data context(s) for rule 'NumberSet' target 'Policy':
         Policy:0
      -> Evaluating rule 'NumberSet' on Policy:0:termNo
         -- Evaluated 'NUMBER_SET' to true. undefined ∈ [1, ∞] with step 1.
      <- Evaluated rule 'NumberSet' on Policy:0:termNo. Evaluation status - APPLIED.
   <- Evaluated rule 'NumberSet' on a total of 1 instances.
   -> Evaluating rule 'ValueList'
      -- Resolved 1 data context(s) for rule 'ValueList' target 'Policy':
         Policy:0
      -> Evaluating rule 'ValueList' on Policy:0:contractTermTypeCd
         -- Evaluated 'VALUE_LIST' to true. Field value '' .
      <- Evaluated rule 'ValueList' on Policy:0:contractTermTypeCd. Evaluation status - APPLIED.
   <- Evaluated rule 'ValueList' on a total of 1 instances.
<- Kraken logs: 'EngineDebugLoggingTest'. Timestamp: Sat, 01 Jan 2022 10:00:00 GMT"
`;
